pkg/apis/rbac/types.go

```diff
@@ -0,0 +1,252 @@
+	Verbs sets.String
+	// AttributeRestrictions will vary depending on what the Authorizer/AuthorizationAttributeBuilder pair supports.
+	// If the Authorizer does not recognize how to handle the AttributeRestrictions, the Authorizer should report an error.
+	AttributeRestrictions runtime.Object
+	// APIGroups is the name of the APIGroup that contains the resources.  If this field is empty, then the core API Group is assumed.
+	// If multiple API groups are specified, any action requested against one of the enumerated resources in any API group will be allowed.
+	APIGroups []string
+	// Resources is a list of resources this rule applies to.  ResourceAll represents all resources.
+	Resources sets.String
+	// ResourceNames is an optional white list of names that the rule applies to.  An empty set means that everything is allowed.
+	ResourceNames sets.String
+	// NonResourceURLs is a set of partial urls that a user should have access to.  *s are allowed, but only as the full, final step in the path
+	// If an action is not a resource API request, then the URL is split on '/' and is checked against the NonResourceURLs to look for a match.
+	NonResourceURLs sets.String
+}
+
+// Role is a logical grouping of PolicyRules that can be referenced as a unit by RoleBindings.
+type Role struct {
+	unversioned.TypeMeta
+	// Standard object's metadata.
+	api.ObjectMeta
+
+	// Rules holds all the PolicyRules for this Role
+	Rules []PolicyRule
+}
+
+// RoleBinding references a Role, but not contain it.  It can reference any Role in the same namespace or in the global namespace.
+// It adds who information via Users and Groups and namespace information by which namespace it exists in.  RoleBindings in a given
+// namespace only have effect in that namespace (excepting the master namespace which has power in all namespaces).
+type RoleBinding struct {
+	unversioned.TypeMeta
+	api.ObjectMeta
+
+	// Subjects hold object references of to authorize with this rule
+	Subjects []api.ObjectReference
+
+	// RoleRef can only reference the current namespace and the global namespace
+	// If the RoleRef cannot be resolved, the Authorizer must return an error.
+	// Since Policy is a singleton, this is sufficient knowledge to locate a role
+	RoleRef api.ObjectReference
+}
+
+// Policy is a object that holds all the Roles for a particular namespace.  There is at most
+// one Policy document per namespace.
+type Policy struct {
+	unversioned.TypeMeta
+	api.ObjectMeta
+
+	// LastModified is the last time that any part of the Policy was created, updated, or deleted
+	LastModified unversioned.Time
+
+	// Roles holds all the Roles held by this Policy, mapped by Role.Name
+	Roles map[string]*Role
+}
+
+// PolicyBinding is a object that holds all the RoleBindings for a particular namespace.  There is
+// one PolicyBinding document per referenced Policy namespace
+type PolicyBinding struct {
```

Comment:
Have you considered how you'll rules for which roles a user is allowed to bind to a given user?  We used this object to give us control over that, but @erictune suggested actually using some sort of a permission on the Role object.  Something like "create" on `roles/binding` could give you the control you need without this wrapper object.

These wrappers do complicate RESTStorage.